### PR TITLE
Adding Terraform resources for Tenancy APIs in GKEHub

### PR DESCRIPTION
**Pull request by sahsagar-google**
Fixes [15167](https://github.com/hashicorp/terraform-provider-google/issues/15167)

If this PR is for Terraform, I acknowledge that I have:

- [X] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [X] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [X] [Generated Terraform providers](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/generate-providers.md), and ran [`make test` and `make lint`](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/run-provider-tests/#run-unit-tests) in the generated providers to ensure it passes unit and linter tests.
- [X] [Ran](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/run-provider-tests.md) relevant acceptance tests using my own Google Cloud project and credentials (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [X] Read [Write release notes](https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
gkehub: added `Namespace`, `Scope`, `MembershipBinding`, `MembershipRBACRoleBinding`, `ScopeRBACRoleBinding` resources
```

### CODE REVIEW

**Comment by sahsagar-google** [comment]
```release-note:enhancement
gkehub: added `Namespace`, `Scope`, `MembershipBinding`, `MembershipRBACRoleBinding`, `ScopeRBACRoleBinding` resources
```

**Review by NickElliot — COMMENTED** [review]
Hey! I've added one small comment that was causing a lint error, but wanted to ask if you could break this PR up into smaller layered ones for each individual resource (or if there are interdependent ones, just those that have to exist together). A five resource review ends up being somewhat unsustainable and likely to result in more review passes than them being separate, resulting it taking longer to reach the provider.

**Review by NickElliot — COMMENTED** [review]
In case you missed the previous comment: https://github.com/GoogleCloudPlatform/magic-modules/pull/8396#pullrequestreview-1538109275

Thanks!

File: mmv1/products/gkehub2/Scope.yaml [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+iam_policy: !ruby/object:Api::Resource::IamPolicy
+  parent_resource_attribute: scope_id
+  exclude: false
+  method_name_separator: ':'
+  import_format:
+    [
+      'projects/{{project}}/locations/global/scopes/{{scope_id}}',
+      '{{scope_id}}',
+    ]
+autogen_async: true
```
[the remark this answers is not in the document]
Comment (sahsagar-google) [reply mmv1/products/gkehub2/Scope.yaml]: Hey Thanks for the comment. I made the change suggested. Also, segregated separate resources and added two more PRs 
https://github.com/GoogleCloudPlatform/magic-modules/pull/8440
https://github.com/GoogleCloudPlatform/magic-modules/pull/8437
Please let me know if this looks better now, Thank you.

File: mmv1/products/gkehub2/Namespace.yaml [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+--- !ruby/object:Api::Resource
+name: 'Namespace'
+base_url: 'projects/{{project}}/locations/global/scopes/{{scope_id}}/namespaces'
```
Comment (NickElliot) [inline mmv1/products/gkehub2/Namespace.yaml:16]: Is this the correct URL location? The API documentation I can find at https://cloud.google.com/anthos/fleet-management/docs/reference/rest/v1beta/projects.locations.namespaces seems to be referencing this without the scopes sub directory.

File: mmv1/products/gkehub2/Namespace.yaml [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+update_url: 'projects/{{project}}/locations/global/scopes/{{scope_id}}/namespaces/{{scope_namespace_id}}'
+self_link: 'projects/{{project}}/locations/global/scopes/{{scope_id}}/namespaces/{{scope_namespace_id}}'
+update_verb: :PATCH
+update_mask: true
+description: |
+  Namespace represents a namespace across the Fleet.
+references: !ruby/object:Api::Resource::ReferenceLinks
+  guides:
+    'Registering a Cluster': 'https://cloud.google.com/anthos/multicluster-management/connect/registering-a-cluster#register_cluster'
+  api: 'https://cloud.google.com/anthos/fleet-management/docs/reference/rest/v1/projects.locations.scopes'
```
Comment (NickElliot) [inline mmv1/products/gkehub2/Namespace.yaml:27]: This needs to reference the API for specifically this resource, the page you linked has no reference to the Namespace resource.

**Review by NickElliot — CHANGES_REQUESTED** [review]
Starting from the top, the namespace test appears to be failing with a 404 which at a glance appears to be due to the URL not being accurate based on what I'm seeing on the documentation. Is there a page I'm not seeing? Or if what I linked in the comment is correct, could you try updating that and running the test locally to see if it passes? Thanks!

**Review by NickElliot — COMMENTED** [review]
still pending requested changes

**Comment by sahsagar-google** [comment]
> Starting from the top, the namespace test appears to be failing with a 404 which at a glance appears to be due to the URL not being accurate based on what I'm seeing on the documentation. Is there a page I'm not seeing? Or if what I linked in the comment is correct, could you try updating that and running the test locally to see if it passes? Thanks!

I updated with the correct documentation. 
The tests were passing locally, but they might have failed because the APIs were still becoming visible at the time but they should be fully visible now, so I am hoping they must pass now. Please lmk if that is not the case. Thank you for bearing with this.

**Comment by sahsagar-google** [comment]
> still pending requested changes

I am rather new to Magic modules. Please let me know if there are more changes required. Thank you

**Review by NickElliot — CHANGES_REQUESTED** [review]
The failing test appears to be the result of the API returning a value using the project id number rather than just its id, which is causing a permadiff. To resolve this a DiffSupressFunc ([example usage](https://github.com/search?q=repo%3AGoogleCloudPlatform%2Fmagic-modules+diff_suppress_func+language%3AYAML&type=code&l=YAML)) will be needed for the field, but thankfully in this case we have a standard one that should hopefully resolve the issue (https://github.com/GoogleCloudPlatform/magic-modules/blob/7aca11d48dee22c24432a4a03455de3bacd79a73/mmv1/third_party/terraform/tpgresource/common_diff_suppress.go.erb#L240C1-L250C2), but if it doesn't a more custom one will be needed.

**Comment by melinath** [comment]
Reassigning review to @ScottSuarez since @NickElliot is not available.

**Comment by melinath** [comment]
Reassigning to @roaks3

**Review by NickElliot — CHANGES_REQUESTED** [review]
The unit tests are failing due to the following: `resource google_gke_hub_namespace: All fields are ForceNew or Computed w/out Optional, Update is superfluous`

i.e. it is suggesting the update_url and related attributes be removed from Namespace.yaml as they server no purpose with the current implementation here -- should "namespaceLabels" or "labels" which are updateable api fields be included in this resource?

**Comment by sahsagar-google** [comment]
> The unit tests are failing due to the following: `resource google_gke_hub_namespace: All fields are ForceNew or Computed w/out Optional, Update is superfluous`
> 
> i.e. it is suggesting the update_url and related attributes be removed from Namespace.yaml as they server no purpose with the current implementation here -- should "namespaceLabels" or "labels" which are updateable api fields be included in this resource?

@NickElliot I plan to update all the resources with labels in a new PR since this is a new change in the proto for me. And since, there is nothing to update in the Namespace, I asked the team that it's okay to make it immutable for now. Please let me know if the tests work now. Thanks!

**Review by NickElliot — CHANGES_REQUESTED** [review]
Could the update test for RBACRoleBinding perform the swap between User and Group? 

Otherwise, a couple changes I'm still seeing needed are:
an update test for Scope, because even though only 'all_memberships' is updateable, the terraform in-place update functionality needs to be verified. Same situation for MembershipBinding with 'scope'.

**Comment by sahsagar-google** [comment]
> Tests passed during RECORDING mode: `TestAccGKEHub2MembershipBinding_gkehubMembershipBindingBasicExample_update`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8396/artifacts/b6981a32-9e85-4d66-8004-c87c9293a1b3/recording/TestAccGKEHub2MembershipBinding_gkehubMembershipBindingBasicExample_update.log)] `TestAccGKEHub2ScopeRBACRoleBinding_gkehubScopeRbacRoleBindingBasicExample_update`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8396/artifacts/b6981a32-9e85-4d66-8004-c87c9293a1b3/recording/TestAccGKEHub2ScopeRBACRoleBinding_gkehubScopeRbacRoleBindingBasicExample_update.log)]
> 
> ##### Rerun these tests in REPLAYING mode to catch issues
> No issues found for passed tests after REPLAYING rerun.
> 
> Tests failed during RECORDING mode: `TestAccGKEHub2Scope_gkehubScopeBasicExample_update`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8396/artifacts/b6981a32-9e85-4d66-8004-c87c9293a1b3/build-log/recording_build/TestAccGKEHub2Scope_gkehubScopeBasicExample_update_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8396/artifacts/b6981a32-9e85-4d66-8004-c87c9293a1b3/recording/TestAccGKEHub2Scope_gkehubScopeBasicExample_update.log)]
> 
> Please fix these to complete your PR. View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8396/artifacts/b6981a32-9e85-4d66-8004-c87c9293a1b3/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-8396/artifacts/b6981a32-9e85-4d66-8004-c87c9293a1b3/recording) for each test

@NickElliot Could you pinpoint the exact error of the Scope update test(TestAccGKEHub2Scope_gkehubScopeBasicExample_update) to me please? 
I looked through the Debug log, but not able to figure out why it failed.

**Review by NickElliot — COMMENTED** [review]
Sorry for the delay, review was sitting in progress before lunch.

`TestAccGKEHub2Scope_gkehubScopeBasicExample_update` is having a failure due to `all_memberships` not maintaining the `true` value when it attempts to refresh the plan, despite returning a success on the PATCH with the true value. It would appear to be a case of go/dont-break-terraform#new-values-returned-for-old-fields causing a permadiff, but if the field is actually immutable or output only then the YAML needs to be changed to match. As you're more familiar with it being on the service team in question, the intended API behavior needs to be clarified.

**Comment by sahsagar-google** [comment]
> Sorry for the delay, review was sitting in progress before lunch.
> 
> `TestAccGKEHub2Scope_gkehubScopeBasicExample_update` is having a permadiff failure due to `all_memberships` not maintaining the `true` value when it attempts to refresh the plan (with the API saying it currently is `false`), despite returning a success on the PATCH with the `true` value. It would appear to be a case of go/dont-break-terraform#new-values-returned-for-old-fields, but if the field is actually immutable or output only then the YAML needs to be changed to match. As you're more familiar with it being on the service team in question, the intended API behavior needs to be clarified.

Thank you for sharing that error. I talked to the team and we feel it should be removed at this point. Sorry for the confusion.

**Review by NickElliot — APPROVED** [review]
LGTM!

**Pull request merged**